### ISSUE TITLE
Comparing structure model runs

### ISSUE THREAD

**Issue by lesserof2weevils** [gbradburd/conStruct#21]
Hi Gideon, 

I'm running into an error while trying to compare model runs using x.validation. If I set parallel and n.nodes to False and Null (respectively), I run into this issue: 
Error in if (args[["parallel"]] & args[["n.nodes"]] == 1) { : 
  argument is of length zero
 and if I set parallel to FALSE and n.node = 1, I run unto this issue: 
Error in check.data.partitions.covmats(args) : 
you have specified an invalid data partition "data" element is not positive definite

I'm not wanting to run this in parallel or on multiple nodes (my dataset isn't large enough to warrant this, so how would I best go about getting this run with parallel and n.nodes set to default values?

Thanks, 
James 

Entire code below:
`my.xvals <- x.validation(train.prop = 0.9,
                         n.reps = 8,
                         K = 1:5,
                         freqs = AiconStruct.data,
                         data.partitions = NULL,
                         geoDist = geoDist,
                         coords = Aipoints,
                         prefix = "mdl",
                         n.iter = 1e3,
                         make.figs = TRUE,
                         save.files = FALSE,
                         parallel = FALSE,
                         n.nodes = NULL)`

**Comment by gbradburd** [comment]
Hi James,

The first issue was a bug in the `check.parallel.args` function, which I've now fixed and pushed to the git.  The `x.validation` function should run fine with `parallel=FALSE` and `n.nodes=NULL` now.  You can install the updated version of the package with the command:

`devtools::install_github("gbradburd/conStruct",build_vignettes=TRUE)`

W/r/t the second issue, please see the thread in https://github.com/gbradburd/conStruct/issues/20#issuecomment-497108469 about dealing with nonpositive-definite issues. If that doesn't resolve it, please attach your allele frequency data as a .txt data table and I'll take a look.

-Gideon

**Comment by lesserof2weevils** [comment]
Hi Gideon, 

Thanks for the fix, that seems to have resolved the first issue. Now when I run with parallel as false, and n.nodes as null, I run into a nonpositive-definite issue. Yet, when I run a spatial analysis with K=3, it runs perfectly fine without running into issues. This suggests to me that all samples sequenced have different genotypes and there shouldn't be any identical pairs of individuals. I also don't have any individuals that cause linear dependancies between samples (to my knowledge)

Cheers, 
James

**Comment by gbradburd** [comment]
Hi James,

Right, so you're getting the non-positive definite error when you try the cross-validation procedure, which works by: 

  1. dividing the data randomly into "training" and "testing" partitions
  2. fitting the model to the training partition
  3. evaluating the likelihood of the testing partition given the model parameterized using the training partition
  4. repeating steps 1-3 some number of times (in your case, `n.reps=8`)

So, even though you're able to run an analysis on the whole dataset just fine (e.g., the spatial model with `K=3`) at least one of the random training or testing data subsets that `x.validation` is trying to create is not positive definite.  Note that this isn't a bug in conStruct - you're having this problem because of a feature of your data. I'd recommend first trying to drop loci or samples that have lots of missing data and seeing if you still get the issue.

**Comment by lesserof2weevils** [comment]
Hi Gideon, 

Thanks for the advice, I'll begin going through my dataset and examining which samples/loci are missing data. Have you observed any trends in which "too much" missing data can become problematic? Conversely, is there are a way to set the training dataset to train.pop=1 to supersede this issue? I understand this will be computationally extensive, but then I imagine it'll be like analysing datasets for both spatial and non-spatial models at increasing K values to compare cross validation scores

Thanks, 
James

**Comment by gbradburd** [comment]
The cutoff at which missing data becomes a problem will vary dataset by dataset, so there's no clear guideline.  Obviously you've worked hard to get those data and you want to include as much of them as possible, so I'd recommend starting by ranking loci (and/or samples) in order of missing-ness, and iteratively dropping them until (hopefully) your positive definite problems go away.

In response to your second question, if the training proportion of the data were set to 1, there would be no independent testing partition, which would make the CV procedure invalid.

hope that helps,
-Gideon

**Comment by lesserof2weevils** [comment]
Hi Gideon, 

I've removed loci with missing values from my dataset (as a broad brush kind of approach), with no NA in the matrix, however, I'm still encountering the issue. 

` my.xvals <- x.validation(train.prop = 0.9,
+                          n.reps = 8,
+                          K = 1:5,
+                          freqs = validconStruct.data,
+                          data.partitions = NULL,
+                          geoDist = geoDist,
+                          coords = Aipoints,
+                          prefix = "mdl",
+                          n.iter = 1e3,
+                          make.figs = TRUE,
+                          save.files = FALSE,
+                          parallel = FALSE,
+                          n.nodes = NULL)
Error in check.data.partitions.covmats(args) : 
you have specified an invalid data partition "data" element that is not positive definite'

Unsure of where to go now; would it be okay for me to email you the .txt file of the matrix for you to take a look at?

Thanks, 
James

**Comment by gbradburd** [comment]
Hi James,

Sure - please either email the allele frequency data matrix you're using (`validconStruct.data`, it looks like?) as a .Robj or attach it as a .txt to this issue thread.  I don't need the coordinates or geographic distance matrix.

**Comment by lesserof2weevils** [comment]
HI Gideon, 

Thanks for taking the time for this. I've uploaded two allele frequency matrices, one with all NAs removed, and the other is one I'd like to analyse. Both datasets return the "Error in check.data.partitions.covmats(args) : 
you have specified an invalid data partition "data" element that is not positive definite" issue.

Thanks, 
James 

[James_NA_absent.txt](https://github.com/gbradburd/conStruct/files/3439875/James_NA_absent.txt)
[James_NA_present.txt](https://github.com/gbradburd/conStruct/files/3439876/James_NA_present.txt)

**Comment by gbradburd** [comment]
Hi James,

In looking at your data, I think the issue is that you don't have many loci in your dataset.  This is a problem for a few reasons.  First, to run cross-validation, you need to have more loci than samples in **each** partition.  You have 93 samples and only 750ish loci, so if we were to run using `train.prop=0.9`, each testing partition would contain only 75 loci, which would make it impossible to evaluate the predictive accuracy of the model on the testing partition.  I hadn't anticipated this problem, so I'll add a check into the cross-validation function to make sure there's an informative error message if this comes up.

The second way the size of the dataset is a problem is in the non-positive-definiteness.  I poked around and don't see any obvious linear dependencies (e.g., Ind3 is exactly intermediate between Inds 1 & 2), which is good.  But, it looks like your samples are partitioned into 3 main genetic groups, with reasonable divergence between, and very low (on average) divergence within each group.  Because you have so few loci, and because within-group divergence is so low, it's not unlikely that a pair of individuals within one of those 3 groups are identical at all the loci randomly included in a testing partition.  This may also interact with some missing data issues.

So, there are a few ways forward.  If it's possible for you to get more data, that'd be the best, but that's probably not super easy to do.  An alternative is to drop some samples and increase the size of the testing partition.  When I drop all samples with a proportion of missing data > 0.03, and run `x.validation()` with `train.prop=0.6`, it seems to work most of the time.  A slightly more complex solution is to "collapse" very closely related individuals into the same sample.  Note that you just have to provide allele frequencies, and that the number of chromosomes in each sample doesn't have to be the same.  So, I'm imagining something like:

1. Identify pairs of individuals that are very closely related
2. Add their allele counts and divide by the sum of their chromosomes (e.g., if they're both diploid, add their diploid genotypes and divide by 4) to get sample frequencies
3. Replace each individual sample being collapsed with the single franken-sample.
4. Run `x.validation` using the resulting dataset.

Hope that's helpful!

**Comment by lesserof2weevils** [comment]
Hi Gideon, 

I've recognised this issue with some other analyses, there is little divergence within groups and they don't like to travel too far from their nests. I'm aiming to get more samples and more populations next field season, but want to assess what we have at the moment. 

I'll try running again with these suggestions, and let you know how the analysis runs.

Thanks for all the help!
James

**Comment by gbradburd** [comment]
Hi James,

Sounds good. I'm going to close the comment, but feel free to open a new issue (or continue to comment on this one) if you hit future snags.  Good luck!

-Gideon
